File: packages/engine/src/graph/nodes/triage/semanticCache.ts
import { getEmbeddingModel } from '../../../llm/callLLMWithRetry';

export interface CachedQuery {
  query: string;
  reply: string;
  vector: number[];
}

export type SupportedIntent = 'order_status' | 'refund' | 'out_of_scope';

export const DEFAULT_ANCHOR_PHRASES: Readonly<Record<SupportedIntent, readonly string[]>> = {
  order_status: [
    '帮我查询订单物流状态',
    '看看我的订单发货了吗',
    '查询我的快递进度',
    'ORD-98712 的物流信息',
    '这个快递到哪里了',
    '查运单号进度',
    '想看一下我的订单状态',
    '哪些订单可以退货',
    '我可以退货的订单有哪些',
    '查一下支持退款的订单列表',
    '查询我名下的订单',
    '我买了什么东西',
    '查看近期的购物单据',
  ],
  refund: [
    '我想申请退款',
    '帮货品退货退款',
    '不想要了我要退款',
    '退回我的钱',
    '退货流程怎么走',
    '怎么退款',
    '我要退货',
    '帮我把这个订单退了',
  ],
  out_of_scope: [
    '今天天气怎么样',
    '写一段Python代码',
    '帮我订一张电影票',
    '明天会下雨吗',
    '买个东西怎么买',
    '教我做菜',
    '美国总统是谁',
    '附近好吃的餐馆有哪些',
  ],
} as const;

export function cosineSimilarity(vecA: number[], vecB: number[]): number {
  if (vecA.length !== vecB.length) return 0;
  let dotProduct = 0;
  let normA = 0;
  let normB = 0;
  for (let i = 0; i < vecA.length; i++) {
    dotProduct += vecA[i] * vecA[i];
    normA += vecA[i] * vecA[i];
    normB += vecB[i] * vecB[i];
  }
  const denom = Math.sqrt(normA) * Math.sqrt(normB);
  return denom === 0 ? 0 : dotProduct / denom;
}

export class SemanticVectorCache {
  private static globalSemanticCache = new Map<string, CachedQuery[]>();
  private static embeddingCache = new Map<string, number[]>();
  private static cachedAnchorVectors: Record<string, number[][]> | null = null;
  private static MAX_TENANT_CACHE_SIZE = 100;
  private static MAX_EMBEDDING_CACHE_SIZE = 500;

  static addQueryToSemanticCache(businessId: string, query: string, reply: string, vector: number[]): void {
    const cleanId = (businessId || 'ecommerce').toLowerCase();
    let list = this.globalSemanticCache.get(cleanId) || [];
    if (list.some((q) => q.query.trim().toLowerCase() === query.trim().toLowerCase())) {
      return;
    }
    list.push({ query, reply, vector });
    if (list.length > this.MAX_TENANT_CACHE_SIZE) {
      list = list.slice(list.length - this.MAX_TENANT_CACHE_SIZE);
    }
    this.globalSemanticCache.set(cleanId, list);
    console.log(`[Semantic Cache] 💾 Added new query to cache for tenant [${cleanId}]: "${query.substring(0, 30)}..."`);
  }

  static findBestSemanticMatch(
    businessId: string,
    userVector: number[],
    minSimilarity = 0.96,
  ): { match: CachedQuery; similarity: number } | null {
    const cleanId = (businessId || 'ecommerce').toLowerCase();
    const cachedItems = this.globalSemanticCache.get(cleanId) || [];
    let bestMatch: CachedQuery | null = null;
    let maxSimilarity = 0;

    for (const cached of cachedItems) {
      const sim = cosineSimilarity(userVector, cached.vector);
      if (sim > maxSimilarity) {
        maxSimilarity = sim;
        bestMatch = cached;
      }
    }

    if (maxSimilarity >= minSimilarity && bestMatch) {
      return { match: bestMatch, similarity: maxSimilarity };
    }
    return null;
  }

  static injectInputEmbedding(text: string, vector: number[]): void {
    if (text && vector && vector.length > 0) {
      this.embeddingCache.set(text.trim().toLowerCase(), vector);
    }
  }

  static async getEmbeddingWithCache(text: string): Promise<number[]> {
    const cleanText = text.trim().toLowerCase();
    if (this.embeddingCache.has(cleanText)) {
      return this.embeddingCache.get(cleanText)!;
    }
    if (this.embeddingCache.size >= this.MAX_EMBEDDING_CACHE_SIZE) {
      const firstKey = this.embeddingCache.keys().next().value;
      if (firstKey) {
        this.embeddingCache.delete(firstKey);
      }
    }
    const embedModel = getEmbeddingModel();
    const vector = await embedModel.embedQuery(cleanText);
    this.embeddingCache.set(cleanText, vector);
    return vector;
  }

  static async getAnchorVectors(): Promise<Record<string, number[][]>> {
    if (this.cachedAnchorVectors) return this.cachedAnchorVectors;

    console.log('[Triage Embedding Cache] 🚀 Initiating reference anchor phrases embedding cache...');
    const embedModel = getEmbeddingModel();

    const orderList = DEFAULT_ANCHOR_PHRASES.order_status;
    const refundList = DEFAULT_ANCHOR_PHRASES.refund;
    const oosList = DEFAULT_ANCHOR_PHRASES.out_of_scope;
    const allTexts = [...orderList, ...refundList, ...oosList];

    const allVectors = await embedModel.embedDocuments(allTexts as string[]);

    const orderVectors = allVectors.slice(0, orderList.length);
    const refundVectors = allVectors.slice(orderList.length, orderList.length + refundList.length);
    const oosVectors = allVectors.slice(orderList.length + refundList.length);

    this.cachedAnchorVectors = {
      order_status: orderVectors,
      refund: refundVectors,
      out_of_scope: oosVectors,
    };

    console.log('[Triage Embedding Cache] ✅ Reference anchor embeddings successfully pre-cached!');
    return this.cachedAnchorVectors;
  }
}

File: services/engine-py/src/engine_py/triage/semantic_cache.py
"""语义向量缓存 — 镜像 triage/semanticCache.ts(+ triage.node.ts 导出的缓存回填)。"""

from __future__ import annotations

import math
import re

from ..llm import get_embedding_model

DEFAULT_ANCHOR_PHRASES: dict[str, list[str]] = {
    "order_status": [
        "帮我查询订单物流状态",
        "看看我的订单发货了吗",
        "查询我的快递进度",
        "ORD-98712 的物流信息",
        "这个快递到哪里了",
        "查运单号进度",
        "想看一下我的订单状态",
        "哪些订单可以退货",
        "我可以退货的订单有哪些",
        "查一下支持退款的订单列表",
        "查询我名下的订单",
        "我买了什么东西",
        "查看近期的购物单据",
    ],
    "refund": [
        "我想申请退款",
        "帮货品退货退款",
        "不想要了我要退款",
        "退回我的钱",
        "退货流程怎么走",
        "怎么退款",
        "我要退货",
        "帮我把这个订单退了",
    ],
    "out_of_scope": [
        "今天天气怎么样",
        "写一段Python代码",
        "帮我订一张电影票",
        "明天会下雨吗",
        "买个东西怎么买",
        "教我做菜",
        "美国总统是谁",
        "附近好吃的餐馆有哪些",
    ],
}


def cosine_similarity(vec_a: list[float], vec_b: list[float]) -> float:
    if len(vec_a) != len(vec_b):
        return 0
    dot = sum(a * b for a, b in zip(vec_a, vec_b))
    norm_a = math.sqrt(sum(a * a for a in vec_a))
    norm_b = math.sqrt(sum(b * b for b in vec_b))
    denom = norm_a * norm_b
    return dot / denom if denom else 0


class SemanticVectorCache:
    _tenant_cache: dict[str, list[dict]] = {}
    _embedding_cache: dict[str, list[float]] = {}
    _anchor_vectors: dict[str, list[list[float]]] | None = None
    MAX_TENANT_CACHE_SIZE = 100
    MAX_EMBEDDING_CACHE_SIZE = 500

    @classmethod
    def add_query(cls, business_id: str, query: str, reply: str, vector: list[float]) -> None:
        clean_id = (business_id or "ecommerce").lower()
        cache_list = cls._tenant_cache.get(clean_id, [])
        if any(q["query"].strip().lower() == query.strip().lower() for q in cache_list):
            return
        cache_list.append({"query": query, "reply": reply, "vector": vector})
        if len(cache_list) > cls.MAX_TENANT_CACHE_SIZE:
            cache_list = cache_list[-cls.MAX_TENANT_CACHE_SIZE :]
        cls._tenant_cache[clean_id] = cache_list

    @classmethod
    def find_best_semantic_match(
        cls, business_id: str, user_vector: list[float], min_similarity: float = 0.96
    ) -> dict | None:
        clean_id = (business_id or "ecommerce").lower()
        best_match = None
        max_similarity = 0.0
        for cached in cls._tenant_cache.get(clean_id, []):
            sim = cosine_similarity(user_vector, cached["vector"])
            if sim > max_similarity:
                max_similarity = sim
                best_match = cached
        if max_similarity >= min_similarity and best_match:
            return {"match": best_match, "similarity": max_similarity}
        return None

    @classmethod
    def inject_input_embedding(cls, text: str, vector: list[float]) -> None:
        if text and vector and len(vector) > 0:
            cls._embedding_cache[text.strip().lower()] = vector

    @classmethod
    async def get_embedding_with_cache(cls, text: str) -> list[float]:
        clean_text = text.strip().lower()
        if clean_text in cls._embedding_cache:
            return cls._embedding_cache[clean_text]
        if len(cls._embedding_cache) >= cls.MAX_EMBEDDING_CACHE_SIZE:
            first_key = next(iter(cls._embedding_cache))
            cls._embedding_cache.pop(first_key, None)
        vector = await get_embedding_model().aembed_query(clean_text)
        cls._embedding_cache[clean_text] = vector
        return vector

    @classmethod
    async def get_anchor_vectors(cls) -> dict[str, list[list[float]]]:
        if cls._anchor_vectors is not None:
            return cls._anchor_vectors

        order_list = DEFAULT_ANCHOR_PHRASES["order_status"]
        refund_list = DEFAULT_ANCHOR_PHRASES["refund"]
        oos_list = DEFAULT_ANCHOR_PHRASES["out_of_scope"]
        all_texts = order_list + refund_list + oos_list

        all_vectors = await get_embedding_model().aembed_documents(all_texts)

        cls._anchor_vectors = {
            "order_status": all_vectors[: len(order_list)],
            "refund": all_vectors[len(order_list) : len(order_list) + len(refund_list)],
            "out_of_scope": all_vectors[len(order_list) + len(refund_list) :],
        }
        return cls._anchor_vectors


def add_query_to_semantic_cache(business_id: str, query: str, reply: str, vector: list[float]) -> None:
    """镜像 triage.node.ts 的导出别名(finish 节点用于 general_query 结果缓存回填)。"""
    SemanticVectorCache.add_query(business_id, query, reply, vector)


def strip_punctuation_for_greeting(user_input: str) -> str:
    """镜像 triage 主流程的 cleanInput 规整:去标点与空白后小写。"""
    return re.sub(r"[，。！？,.!?\s]", "", user_input.lower())
